packages/tools/src/lib/build-cache/common.ts

```diff
@@ -10,28 +10,79 @@ export type SupportedRemoteCacheProviders = 'github-actions';
 
 export type RemoteArtifact = {
   name: string;
-  downloadUrl: string;
+  url: string;
+  id?: string;
 };
 
 export type LocalArtifact = {
   name: string;
   path: string;
 };
 
+/**
+ * Interface for implementing remote build cache providers.
+ * Remote cache providers allow storing and retrieving native build artifacts (e.g. APK, IPA)
+ * from remote storage like S3, GitHub Artifacts etc.
+ */
 export interface RemoteBuildCache {
+  /** Unique identifier for this cache provider, will be displayed in logs */
   name: string;
-  query({
+
+  /**
+   * List available artifacts matching the given name pattern
+   * @param artifactName - Passed after fingerprinting the build, e.g. `rnef-android-debug-1234567890` for android in debug variant
+   * @param limit - Optional maximum number of artifacts to return
+   * @returns Array of matching remote artifacts, or null if none found
+   */
+  list({
     artifactName,
+    limit,
   }: {
-    artifactName: string;
-  }): Promise<RemoteArtifact | null>;
+    artifactName: string | undefined;
+    limit?: number;
+  }): Promise<RemoteArtifact[] | null>;
+
+  /**
+   * Download a remote artifact to local storage
+   * @param artifact - Remote artifact to download, as returned by `list` method
+   * @param loader - Optional progress indicator
+   * @returns Local artifact info after download
+   */
   download({
     artifact,
     loader,
   }: {
     artifact: RemoteArtifact;
-    loader: ReturnType<typeof spinner>;
+    loader?: ReturnType<typeof spinner>;
   }): Promise<LocalArtifact>;
```

Comment:
isn't the local artifact path determined by rnef (the `.rnef` folder)?

imo it may be better if this function returns a stream or buffer containing the file instead. so rnef can store it in its cache folder (and that folder could be changed by changing rnef config without requiring cache implementations to be aware of it)

Reply:
So for now I've changed it so that download now gets the `targetURL` which is the directory controlled by RNEF that you can download the files to. Not sure about a stream yet. I like the idea, as we could move the download progress code inside. But currently for example when downloading iOS .app file from GitHub, it's compressed so the download handler needs to uncompress it: https://github.com/callstack/rnef/blob/e8c9da7cda1f90e0cc7802dc2ace16f0c883ecca/packages/tools/src/lib/build-cache/github/GitHubBuildCache.ts#L91

I'm happy to revisit this once we implement more providers